src/iostream.c

```diff
@@ -531,15 +551,23 @@ Obj FuncREAD_IOSTREAM( Obj self, Obj stream, Obj len )
   Int ret;
   Obj string;
   string = NEW_STRING(INT_INTOBJ(len));
-  while (!PtyIOStreams[pty].inuse)
-    pty = INT_INTOBJ(ErrorReturnObj("IOSTREAM %d is not in use",pty,0L,
-                                    "you can replace stream number <num> via 'return <num>;'"));
+  HashLock(PtyIOStreams);
+  if (!PtyIOStreams[pty].inuse) {
+    HashUnlock(PtyIOStreams);
+    ErrorMayQuit("IOSTREAM %d is not in use",pty,0L);
+    return Fail;
+  }
+  
   /* HandleChildStatusChanges(pty);   Omit this to allow picking up "trailing" bytes*/
   ret = ReadFromPty2(pty, CSTR_STRING(string), INT_INTOBJ(len), 1);
-  if (ret == -1)
+  if (ret == -1) {
+    HashUnlock(PtyIOStreams);
     return Fail;
+  }
   SET_LEN_STRING(string, ret);
```

Comment:
OK -- so, were is this documented (and out of curiosity: why is it so?). This strikes me as very important information (sadly, most HPC-GAP kernel code seems to undocumented, with no hints about requirements or prerequisites or gotachas or about what they do... :/).